### PR TITLE
Remove Spring Boot dependency management

### PR DESCRIPTION
**Pull request by cuijulian**
Depends on https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1086

### CODE REVIEW

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by MarianMacik** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by MarianMacik** [comment]
jenkins execute springboot build

File: kie-spring-boot/pom.xml [hunk 1]
```diff
@@ -44,24 +42,6 @@
...
-	<groupId>org.springframework.boot</groupId>
-	<artifactId>spring-boot-starter</artifactId>
-	<version>${version.org.spring-boot}</version>
-	<exclusions>
-	    <exclusion>
-	      <groupId>org.apache.logging.log4j</groupId>
-              <artifactId>log4j-api</artifactId>
-            </exclusion>
-	  </exclusions>
-      </dependency>
```
Comment (MarianMacik) [inline kie-spring-boot/pom.xml]: I am a little bit worried about this exclusion which is not materialized here currently. Maybe we should leave it here as it was.

Comment (cuijulian) [reply kie-spring-boot/pom.xml]: Good point.

**Comment by cuijulian** [comment]
jenkins retest this

**Comment by MarianMacik** [comment]
@cuijulian So I checked why the build fails on Appformer repository but I haven't found anything. For some reason we have a compilation error in the uberfire-structure-backend module:

`[ERROR] /home/jenkins/workspace/KIE/master/pullrequest/droolsjbpm-integration-springboot-pullrequests/upstream-repos/appformer/uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/pom/PomJsonReaderDefault.java:[38,13] cannot find symbol`

Appformer is built after `droolsjbpm-build-bootstrap`, so there is some influence from the `droolsjbpm-build-bootstrap` on the Appformer repo.

**Comment by cuijulian** [comment]
> @cuijulian So I checked why the build fails on Appformer repository but I haven't found anything. For some reason we have a compilation error in the uberfire-structure-backend module:
> 
> `[ERROR] /home/jenkins/workspace/KIE/master/pullrequest/droolsjbpm-integration-springboot-pullrequests/upstream-repos/appformer/uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/pom/PomJsonReaderDefault.java:[38,13] cannot find symbol`
> 
> Appformer is built after `droolsjbpm-build-bootstrap`, so there is some influence from the `droolsjbpm-build-bootstrap` on the Appformer repo.

@MarianMacik I can take a look at the Appformer repo. This is strange; even though Appformer inherits from `kie-parent` it doesn't use any Spring dependencies. I suspect the newer version of Spring Boot overrides other dependency versions in Appformer.

**Comment by cuijulian** [comment]
jenkins retest this

**Comment by cuijulian** [comment]
jenkins execute full downstream build

**Comment by cuijulian** [comment]
@MarianMacik I resolved the Appformer compilation issue (it was what I suspected, adding Spring Boot 2.2 to `kie-parent` introduced transitive dependencies that overrode required dependencies in other repos, such as Appformer). But the downstream build exposed yet another problem: a duplicate dependency class in `kie-camel`. I will push a fix tomorrow.

**Comment by MarianMacik** [comment]
@cuijulian Thank you. I thought that `javax.json.JsonObject` is a part of JDK, but it is from Java EE. Now it makes sense why it failed. In case you have any issues excluding `kie-camel` duplicate classes, ping me and I will look at it as well.

**Comment by cuijulian** [comment]
@MarianMacik Sorry, one correction. The duplicate classes were found in `droolsjbpm-bpms-distribution` and `droolsjbpm-brms-distribution` according to the Jenkins build:

`06:05:09 Duplicate classes found:
06:05:09 
06:05:09   Found in:
06:05:09     org.slf4j:jcl-over-slf4j:jar:1.7.26:compile
06:05:09     org.springframework:spring-jcl:jar:5.2.0.RELEASE:compile
06:05:09   Duplicate classes:
06:05:09     org/apache/commons/logging/Log.class
06:05:09     org/apache/commons/logging/impl/NoOpLog.class
06:05:09     org/apache/commons/logging/impl/SimpleLog.class
06:05:09     org/apache/commons/logging/LogFactory.class`

However, the Jenkins build result is inconsistent with my local build, which works fine. The [POM](https://github.com/kiegroup/droolsjbpm-integration/blob/master/kie-spring/pom.xml#L144) in `kie-spring` clearly excludes the commons-logging dependency which should solve this issue, but for some reason the build still fails. Has this happened before, and what should I do?

**Comment by MarianMacik** [comment]
@cuijulian In fact, duplicated classes were found in these modules:

* kie-camel
* kie-spring-test
* jbpm-spring-boot-starter-basic
* kie-server-spring-boot-autoconfiguration
* droolsjbpm-bpms-distribution
* droolsjbpm-brms-distribution

So you were 50% right :) When you look at the console output, at the end there are all failures listed after the `BUILD FAILURE`. The reason why it's better to look at all the failures at the bottom of the build log is that the Maven build is run using the `-fae (fail at end)` option which will continue building modules even after the failure. That's the reason why not only `BPMS distribution` and `BRMS distribution` have `FAILURE` listed but others as well.

With kie-camel I had to exclude the dependency also from camel-cxf https://github.com/kiegroup/droolsjbpm-integration/pull/1931/files#diff-d6743a971a2d28ef7caa84bf53ee4139R166 even though from the Maven perspective it is not needed - dependency tree is without the spring-jcl artifact even without that exclusion. But the enforcer plugin was still showing issues - probably there is either a bug in the enforcer plugin or OSGI bundles cause it troubles for some reason. I have also included some JakartaEE artifacts even though the enforcer didn't find it but when I opened them they contained classes already present on our classpath in JavaEE packages.

And regarding different results on a local machine and on Jenkins, did you try to run your local build with the `-Dfull` option? Maybe that caused the discrepancy.

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by cuijulian** [comment]
jenkins retest this

**Comment by cuijulian** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
@mareknovotny 

```
The following artifacts could not be resolved:
io.netty:netty-buffer:jar:3.10.6.Final,
io.netty:netty-transport:jar:3.10.6.Final,
io.netty:netty-handler:jar:3.10.6.Final,
io.netty:netty-transport-native-epoll:jar:linux-x86_64:3.10.6.Final,
io.netty:netty-transport-native-kqueue:jar:osx-x86_64:3.10.6.Final,
io.netty:netty-codec-http:jar:3.10.6.Final: Could not find artifact io.netty:netty-buffer:jar:3.10.6.Final in mirror-central
```
I guess it is connected with the change in the droolsjbpm-build-bootstrap PR which changed the version of netty?

**Comment by mareknovotny** [comment]
@MarianMacik yes, this is not ok. I already added to your comment there that this is not ok

**Comment by cuijulian** [comment]
jenkins execute full downstream build

**Comment by cuijulian** [comment]
jenkins retest this

**Comment by cuijulian** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
Spring Boot job failed to compile because of kiegroup/kie-jenkins-scripts#638. It should be fixed now.

**Comment by MarianMacik** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by cuijulian** [comment]
jenkins retest this

**Comment by cuijulian** [comment]
jenkins execute springboot build

**Comment by cuijulian** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
@cuijulian Woohoo! Everything is green!

File: kie-camel/pom.xml [hunk 2]
```diff
@@ -114,7 +114,6 @@
     <dependency>
       <groupId>org.kie</groupId>
       <artifactId>kie-spring</artifactId>
-      <scope>test</scope>
```
Comment (MarianMacik) [inline kie-camel/pom.xml]: I tried `mvn clean install -DskipTests -Dfull` with the test scope and it works. Can you please put it back to test? Otherwise we would have many artifacts on compile path which are not necessary.

Comment (cuijulian) [reply kie-camel/pom.xml]: @MarianMacik It doesn't work for me. When I run `mvn clean install -DskipTests -Dfull` with the test scope the `maven-enforcer-plugin` complains about duplicate classes caused by `spring-jcl`. This happens even though I clearly exclude that dependency from `kie-spring`. Changing the scope back to compile fixes this. Could this be a bug in the enforcer in test scope?

Also, does it work when you run `mvn clean install -Dfull`? Don't we need to run tests to verify before merging?

Comment (cuijulian) [reply kie-camel/pom.xml]: I'll try adding the test scope back and see what Jenkins says. Perhaps it's an issue with my local machine.

**Comment by cuijulian** [comment]
jenkins retest this

**Comment by cuijulian** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
@cuijulian As I said here: https://github.com/kiegroup/droolsjbpm-integration/pull/1931#issuecomment-552117910 

> With kie-camel I had to exclude the dependency also from camel-cxf https://github.com/kiegroup/droolsjbpm-integration/pull/1931/files#diff-d6743a971a2d28ef7caa84bf53ee4139R166 even though from the Maven perspective it is not needed - dependency tree is without the spring-jcl artifact even without that exclusion. But the enforcer plugin was still showing issues - probably there is either a bug in the enforcer plugin or OSGI bundles cause it troubles for some reason.

So I just did it again. I think the issue is that because kie-spring is test dependency, enforcer somehow excludes just the test scope of jcl? And then will find again another jcl dependency coming from camel-cxf, this time with a normal compile scope. But dependency tree of course doesn't show any jcl dependency coming from camel-cxf because it has already been pulled from kie-spring.

The other option which worked was to declare kie-spring test dependency after camel-cxf so camel-cxf pulls jcl and excludes it and enforcer was ok with that.

So apparently enforcer works differently than Maven resolution does and it seems to only respect exclusions from non-test dependencies? But at the same time this works with other projects without an issue, so maybe this is a corner case. I even tried latest Maven Enforcer plugin (3.0.0-M3 version) and it didn't help. Now it should be OK.

@mareknovotny just FYI ^^^ that Enforcer plugin may be not that reliable as we thought.

**Comment by MarianMacik** [comment]
jenkins execute springboot build

**Comment by MarianMacik** [comment]
test failure not related - failing also for other PRs

**Comment by cuijulian** [comment]
@MarianMacik could you take a look at this PR when you get a chance: https://github.com/kiegroup/jbpm-work-items/pull/114#issuecomment-558209008

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by MarianMacik** [comment]
@cuijulian Green!

**Comment by cuijulian** [comment]
> @cuijulian Green!

Thanks a lot for your help! Fingers crossed that the full downstream build passes.
